pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+	if err == nil {
+		return nil // file already exists, exit right away.
+	} else if !os.IsNotExist(err) {
+		return err // legitimate error, propagate it.
+	}
+
+	// If we got here, the README.md doesn't already exist -- write out a stock one.
+	w, err := tools.NewGenWriter(tfgen, rf)
+	if err != nil {
+		return err
+	}
+	defer contract.IgnoreClose(w)
+
+	downstreamLicense := g.info.GetTFProviderLicense()
+	licenseTypeURL := getLicenseTypeURL(downstreamLicense)
+	w.Writefmtln(standardDocReadme, g.pkg, g.info.Name, g.info.GetGitHubOrg(), downstreamLicense, licenseTypeURL)
+	return nil
+}
+
+// emitUtilities emits a utilities file for submodules to consume.
+func (g *csharpGenerator) emitUtilities(mod *module) (string, error) {
+	contract.Require(mod.root(), "mod.root()")
+
+	// Open the utilities.ts file for this module and ready it for writing.
+	w, utilities, err := g.openWriter(mod, "Utilities.cs")
+	if err != nil {
+		return "", err
+	}
+	defer contract.IgnoreClose(w)
+
+	// Strip any 'v' off of the version.
+	version := g.info.Version
+	if len(version) > 0 && version[0] == 'v' {
+		version = version[1:]
+	}
+
+	var buf bytes.Buffer
+	err = csharpUtilitiesTemplate.Execute(&buf, csharpUtilitiesTemplateContext{
+		Namespace: g.moduleNamespace(mod),
+		ClassName: "Utilities",
+		Version:   version,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	w.WriteString(buf.String())
+	return utilities, nil
+}
+
+// emitModuleMember emits the given member, and returns the module file that it was emitted into (if any).
+func (g *csharpGenerator) emitModuleMember(
+	mod *module, member moduleMember, nested *csharpNestedTypes) (string, error) {
+
+	glog.V(3).Infof("emitModuleMember(%s, %s)", mod, member.Name())
+
+	switch t := member.(type) {
+	case *resourceType:
+		return g.emitResourceType(mod, t, nested)
+	case *resourceFunc:
+		return g.emitResourceFunc(mod, t, nested)
+	case *variable:
+		contract.Assertf(mod.config(),
+			"only expected top-level variables in config module (%s is not one)", mod.name)
+		// skip the variable, we will process it later.
+		return "", nil
+	case *overlayFile:
+		return g.emitOverlay(mod, t)
+	default:
+		contract.Failf("unexpected member type: %v", reflect.TypeOf(member))
+		return "", nil
+	}
+}
+
+func (g *csharpGenerator) emitConfigVariableType(w *tools.GenWriter, typ *propertyType) {
+	contract.Assert(typ.kind == kindObject)
+
+	w.Writefmtln("")
+
+	// Open the class.
+	w.Writefmtln("    public class %s", typ.name)
+	w.Writefmtln("    {")
+
+	// Generate each output field.
+	for _, prop := range typ.properties {
+		propertyName := csharpPropertyName(prop)
+		propertyType := csharpPropertyType(prop, false /*wrapInput*/)
+
+		initializer := ""
+		if !prop.optional() {
+			initializer = " = null!;"
+		}
+
+		emitCSharpPropertyDocComment(w, prop, "        ")
+		w.Writefmtln("        public %s %s { get; set; }%s", propertyType, propertyName, initializer)
+	}
+
+	// Close the class.
+	w.Writefmtln("    }")
```

Comment:
aside, for hte future: in roslyn i updated our generator to use this pattern:

https://github.com/dotnet/roslyn/blob/73bdf2022b69ff9b7e7c8c8b4dde747d48d5aaa7/src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceWriter.cs#L91-L99

Specifically, all the code spitting out the final lang code doesn't need to worry about indentation.  it just says `Open/CloseBlock`.   That way you can move things around, and have helper functions that work anywhere, and all the indentation is always right because it's not being manually hardcoded in.

It's a simple pattern, but works quite well, and ensures a nice consistent output.

Reply:
Yeah, we do this in `tf2pulumi`. I'm not going to refactor it any time soon, though--it's not really worth the trouble at the moment.